Sources/MapboxNavigation/SpriteRepository.swift

```diff
@@ -0,0 +1,147 @@
+            dispatchGroup.leave()
+        }
+        
+        dispatchGroup.notify(queue: .main) {
+            self.styleURI = styleURI
+            self.baseURL = baseURL
+            completion()
+        }
+    }
+    
+    func spriteURL(isImage: Bool, baseURL: URL, styleID: String) -> URL? {
+        guard var urlComponent = URLComponents(url: baseURL, resolvingAgainstBaseURL: false),
+              let accessToken = NavigationSettings.shared.directions.credentials.accessToken else { return nil }
+        
+        let requestTyep = isImage ? "/sprite@2x.png" : "/sprite@2x"
+        urlComponent.path += styleID
+        urlComponent.path += requestTyep
+        urlComponent.queryItems = [URLQueryItem(name: "access_token", value: accessToken)]
+        return urlComponent.url
+    }
+    
+    func downloadMetadata(_ metadataURL: URL, completion: @escaping (Data?) -> Void) {
+        metadataCache.clearMemory()
+        
+        let _ = imageDownloader.downloadImage(with: metadataURL, completion: { [weak self] (_, data, error)  in
+            guard let strongSelf = self, let data = data else {
+                completion(nil)
+                return
+            }
+
+            guard error == nil else {
+                completion(data)
+                return
+            }
+
+            strongSelf.metadataCache.store(data)
+            completion(data)
+        })
+    }
+    
+    func downLoadSprite(_ spriteURL: URL, completion: @escaping (UIImage?) -> Void) {
+        imageCache.clearMemory()
+        
+        let _ = imageDownloader.downloadImage(with: spriteURL, completion: { [weak self] (image, data, error) in
+            guard let strongSelf = self, let image = image else {
+                completion(nil)
+                return
+            }
+
+            guard error == nil else {
+                completion(image)
+                return
+            }
+
+            strongSelf.imageCache.store(image, forKey: "Sprite", toDisk: false, completion: {
+                completion(image)
+            })
+        })
+    }
+    
+    func getShield(displayRef: String, name: String) -> UIImage? {
+        let iconLeght = (displayRef.count < 2 ) ? 2 : displayRef.count
+        let metadataName = name + "-\(iconLeght)"
+        
+        guard let spriteImage = imageCache.image(forKey: "Sprite"),
+              let spriteMetaData = metadataCache.spriteMetaData(forKey: metadataName) else { return nil }
+        
+        let shieldRect = CGRect(x: spriteMetaData.x, y: spriteMetaData.y, width: spriteMetaData.width, height: spriteMetaData.height)
+        if let croppedCGIImage = spriteImage.cgImage?.cropping(to: shieldRect) {
+            return UIImage(cgImage: croppedCGIImage)
+        }
+        
+        return nil
+    }
+    
+    func getLegacyShield(imageBaseUrl: String, completion: @escaping (UIImage?) -> Void) {
+        guard let requestURL = URL(string: imageBaseUrl + "@2x.png") else {
```

Comment:
Do we need to hardcode 2× resolution here, or can we make it match the screen resolution like before? Also, is there any chance of reusing the existing URL calculation code from `VisualInstruction.Component.ImageRepresentation` in MapboxDirections?

Reply:
Yes. In fact, it's a good idea to use the `ImageRepresentation` for the `SpriteRepository`, it combines the `shield` and  `imageBaseUrl` , and could be used to download both the shield and legacy shield at the same time and store it. Thanks for the suggestion, I'll update this part.